src/libsync/propagatedownload.cpp

```diff
@@ -631,6 +673,108 @@ void PropagateDownloadFile::slotGetFinished()
     validator->start(_tmpFile.fileName(), checksumHeader);
 }
 
+void PropagateDownloadFile::slotZsyncGetFinished()
+{
+    propagator()->_activeJobList.removeOne(this);
+
+    GETFileZsyncJob *job = qobject_cast<GETFileZsyncJob *>(sender());
+    ASSERT(job);
+
+    SyncFileItem::Status status = job->errorStatus();
+    if (status != SyncFileItem::NoStatus) {
+        done(status, job->errorString());
+        return;
+    }
+
+    if (!job->reply()) {
+        downloadFinished();
+        return;
+    }
+
+    QNetworkReply::NetworkError err = job->reply()->error();
+    if (err != QNetworkReply::NoError) {
+        _item->_httpErrorCode = job->reply()->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
+
+        // If we sent a 'Range' header and get 416 back, we want to retry
+        // without the header.
+        const bool badRangeHeader = _item->_httpErrorCode == 416;
+        if (badRangeHeader) {
+            qCWarning(lcPropagateDownload) << "server replied 416 to our range request, trying again without";
+            propagator()->_anotherSyncNeeded = true;
+        }
+
+        // Getting a 404 probably means that the file was deleted on the server.
+        const bool fileNotFound = _item->_httpErrorCode == 404;
+        if (fileNotFound) {
+            qCWarning(lcPropagateDownload) << "server replied 404, assuming file was deleted";
+        }
+
+        // Don't keep the temporary file if it is empty or we
+        // used a bad range header or the file's not on the server anymore.
+        if (_tmpFile.size() == 0 || badRangeHeader || fileNotFound) {
+            _tmpFile.close();
+            FileSystem::remove(_tmpFile.fileName());
+            propagator()->_journal->setDownloadInfo(_item->_file, SyncJournalDb::DownloadInfo());
+        }
+
+        // This gives a custom QNAM (by the user of libowncloudsync) to abort() a QNetworkReply in its metaDataChanged() slot and
+        // set a custom error string to make this a soft error. In contrast to the default hard error this won't bring down
+        // the whole sync and allows for a custom error message.
+        QNetworkReply *reply = job->reply();
```

Comment:
Unfortunate large chunk of duplicated code. Get rid of it if at all possible. What are the differences between GetFinished and ZsyncFinished -- can we arrange such that there aren't any?

Reply:
Unfortunate indeed, `slotGetZsyncFinished` is gone now.